cosmwasm/packages/sgx-vm/src/enclave.rs

```diff
@@ -169,29 +128,82 @@ pub fn configure_enclave(config: EnclaveRuntimeConfig) -> SgxResult<()> {
-    let ptr = buffer.as_ptr();
-    let len = buffer.len();
-    let mut enclave_buffer = EnclaveBuffer::default();
-
-    let enclave_id = get_enclave()
-        .expect("If we got here, surely the enclave has been loaded")
-        .expect("If we got here, surely we are the thread that holds the enclave")
-        .geteid();
-
-    trace!(
-        target: module_path!(),
-        "allocate_enclave_buffer() called with len: {:?} enclave_id: {:?}",
-        len,
-        enclave_id,
-    );
-
-    match unsafe { imports::ecall_allocate(enclave_id, &mut enclave_buffer, ptr, len) } {
-        sgx_status_t::SGX_SUCCESS => Ok(enclave_buffer),
-        failure_status => Err(failure_status),
+/// This const determines how many seconds we wait when trying to get access to the enclave
+/// before giving up.
+const ENCLAVE_LOCK_TIMEOUT: u64 = 6 * 5;
+const TCS_NUM: u8 = 16;
+lazy_static! {
+    static ref QUERY_DOORBELL: Doorbell = Doorbell::new(TCS_NUM);
+}
+
+struct Doorbell {
+    condvar: Condvar,
+    /// Amount of tasks allowed to use the enclave at the same time.
+    count: Mutex<u8>,
+}
+
+impl Doorbell {
+    fn new(count: u8) -> Self {
+        Self {
+            condvar: Condvar::new(),
+            count: Mutex::new(count),
+        }
+    }
+
+    fn wait_for(&'static self, duration: Duration, recursive: bool) -> Option<EnclaveQueryToken> {
+        // eprintln!("Query Token creation. recursive: {}", recursive);
+        if !recursive {
+            let mut count = self.count.lock();
+            // eprintln!(
+            //     "The current count of tasks is {}/{}, attempting to increase.",
+            //     TCS_NUM - *count,
+            //     TCS_NUM
+            // );
+            if *count == 0 {
+                // try to wait for other tasks to complete
+                let wait = self.condvar.wait_for(&mut count, duration);
+                // double check that the count is nonzero, so there's an available slot in the enclave.
+                if wait.timed_out() || *count == 0 {
+                    return None;
+                }
+            }
+            *count -= 1;
+        }
+        Some(EnclaveQueryToken::new(self, recursive))
+    }
+}
+
+pub struct EnclaveQueryToken {
+    doorbell: &'static Doorbell,
+    recursive: bool,
+}
+
+impl EnclaveQueryToken {
+    fn new(doorbell: &'static Doorbell, recursive: bool) -> Self {
+        Self {
+            doorbell,
+            recursive,
+        }
+    }
+}
+
+impl Drop for EnclaveQueryToken {
+    fn drop(&mut self) {
+        // eprintln!("Query Token destruction. recursive: {}", self.recursive);
+        if !self.recursive {
+            let mut count = self.doorbell.count.lock();
+            // eprintln!(
+            //     "The current count of tasks is {}/{}, attempting to decrease.",
+            //     TCS_NUM - *count,
+            //     TCS_NUM
+            // );
+            *count += 1;
```

Comment:
Shouldn't it be `-= 1` ?

Reply:
No, the count represents the amount of "free slots" that are available in the enclave runtime. So when the query token is destroyed, we want to release the slot, by saying there's now one more available.

Reply:
Ahh right right, my bad